Make `offsetGridLines` behavior consistent and add new `offset` option to scales

nagix opened the pull request:
[Document](http://www.chartjs.org/docs/latest/charts/bar.html#offsetgridlines) says

> ### offsetGridLines
> If true, the bars for a particular data point fall between the grid lines. If false, the grid line will go right down the middle of the bars.

However, the current implementation doesn't change the positions of grid lines, and bars stay between the grid lines regardless of whether it is true or false. Instead, only the positions of labels are shifted.

When a chart has mixed data types (e.g. bar and line) and `offsetGridLines` is `false`, the grid line goes right down the middle of the bars as the document explains. This seems to me the correct behavior even for non-mixed charts when `offsetGridLines` is `false`.

Another problem is that the right most bar is completely hidden (in case of a bar chart) or bars at the both edges are hidden partially (in case of a mixed chart) when `offsetGridLines` is `false`. Calculation of the bar width is also incorrect.

<img width="397" alt="screen shot 2017-07-19 at 6 31 18 pm" src="https://user-images.githubusercontent.com/723188/28362816-8e810530-6cb0-11e7-928a-c83067e081cb.png">

In case of time scale charts, `offsetGridLines` has no effect, and bars are drawn as if it were set to `true`. It makes more sense to me that the center of a bar is positioned at the time that the data point refers as the default behavior. The behavior of the `offsetGridLines` option should be consistent across different scales.

The other issue is that the width of bars and padding are calculated based on the number of the ticks. If the number of the samples are fewer or more than the number of the ticks, unnecessary margin or overlaps will appear.

<img width="393" alt="screen shot 2017-07-22 at 12 03 47 am" src="https://user-images.githubusercontent.com/723188/28471940-c2a275d6-6e79-11e7-8f60-88fb37ea7933.png">

This PR contains the following changes to resolve the issues above.

- Remove `chart.isCombo`
- Remove the `includeOffset` argument from `getPixelForValue()` and `getPixelForTick()`
- Add a new `offset` option to scales to add extra space (half tick interval) at edges
- Bar controller automatically calculates the bar width to avoid overlaps
- When `offsetGridLines` is true, grid lines move to the left by one half of the tick interval, and labels don't move
- Add tests to the scales and the bar controller

The following images compare the current version (on the left) and the proposed version (on the right). 

<img width="838" alt="screen shot 2017-08-12 at 3 09 57 am" src="https://user-images.githubusercontent.com/723188/29228403-27c8d21e-7f0c-11e7-82a5-5200e696a953.png">

Below are charts with time scales.

<img width="835" alt="screen shot 2017-08-12 at 3 10 34 am" src="https://user-images.githubusercontent.com/723188/29228413-34c58372-7f0c-11e7-9a87-c7ff14275baa.png">

<img width="822" alt="screen shot 2017-08-12 at 3 11 06 am" src="https://user-images.githubusercontent.com/723188/29228452-57eff0bc-7f0c-11e7-9931-7794e4cdb369.png">

Below are examples of horizontal bar charts.

<img width="841" alt="screen shot 2017-08-12 at 3 11 47 am" src="https://user-images.githubusercontent.com/723188/29228462-639f5bf0-7f0c-11e7-94af-27fc5bb1342f.png">

Below are examples of non-linear time scales.

<img width="934" alt="screen shot 2017-08-12 at 3 22 14 am" src="https://user-images.githubusercontent.com/723188/29228772-c014e5ca-7f0d-11e7-98ee-3ed2a7b33bdb.png">

Fixes #2323
Fixes #2415
Fixes #3263
Fixes #3297
Fixes #3579 
Fixes #3613
Fixes #3682
Fixes #3913
Fixes #4051
Fixes #4343
Fixes #4471
Fixes #4560

etimberg commented:
@nagix thanks for the incredibly detailed PR notes. I will try and look at the code soon.

benmccann commented on src/scales/scale.category.js:
```diff
@@ -46,6 +46,11 @@ module.exports = function(Chart) {
 			// If we are viewing some subset of labels, slice the original array
 			me.ticks = (me.minIndex === 0 && me.maxIndex === labels.length - 1) ? labels : labels.slice(me.minIndex, me.maxIndex + 1);
+
+			// Disable zero line if a chaer has bars and offsetGridLines is false
```
typo `chaer`

benmccann commented on src/scales/scale.time.js:
```diff
@@ -281,6 +281,29 @@ module.exports = function(Chart) {
 				max = ticks.length ? ticks[ticks.length - 1] : max;
 			}
 
+			// If the chart has bars, shift min and max based on the sample size
```
I think you'll need a different logic here if `mode: series`

`ticks.mode` (`'linear'`(default)|`series`): `series` displays ticks at the same distance from each other, whatever the time value they represent, while `linear` displays them linearly in time: the distance between each tick represent the amount of time between their time values.

nagix replied:
Good point! As each tick interval can have different value within the same scale, the sample size of bars can also very. So, I removed the ruler code from controller.bar.js, and made calculateBarIndexPixels() call getBarPixelsForIndex() to get the left and right edge of the bar for a specified index. I also modified the calculation logic in scale.time.js to support `ticks.mode` and `ticks.source`.

benmccann commented on src/scales/scale.time.js:
```diff
@@ -281,6 +281,29 @@ module.exports = function(Chart) {
+			// If the chart has bars, shift min and max based on the sample size
+			if (me.options.gridLines.offsetGridLines || me.chart.hasBar) {
+				var sampleSizeMs = moment.duration(me.options.sampleSize || 0).asMilliseconds();
+				var offsetGridLines = me.options.gridLines.offsetGridLines;
```
if you set this variable outside the `if` then you can use it in the `if` check

benmccann commented on docs/charts/bar.md:
````diff
@@ -116,6 +117,13 @@ options = {
 ```
 
+### sampleSize
+If specified, the width for each sample will be calcurated based on this value. If not specified, it will be the available width (the space between the gridlines). This only works if `barThickness` is not set.
````
calcurated -> calculated

nagix commented:
Below are the time scale charts with `gridLines.offsetGridLines = true`, `tick.mode = linear | series` and `tick.source = auto | labels`. It looks ok now. There are overlaps when `gridLines.offsetGridLines` is `false`, but that problem cannot be solved when tick time and sample time are not aligned, and the fixed sample size (`sampleSize` option) should be used in that case.

<img width="375" alt="screen shot 2017-07-24 at 2 30 44 am" src="https://user-images.githubusercontent.com/723188/28501103-7621088c-7007-11e7-92b8-2451a4bc71a5.png">

benmccann commented:
@nagix looks like there's still a rebase needed

nagix commented:
Just resolved conflicts.

benmccann commented:
I think there's a bug here with `mode: series` and `offsetGridLines: false`. See the screenshot below. It causes the bars to be different widths.

![screenshot from 2017-07-23 23-07-27](https://user-images.githubusercontent.com/322311/28510371-e3445f78-6ffc-11e7-8cf5-a49541d4f567.png)

I tested with an updated version of the financial chart pointed at a locally built version of chart.js with `npm link`: https://github.com/chartjs/chartjs-chart-financial/tree/chartjs-2.7

I sent a PR to add a similar sample to this repo https://github.com/chartjs/Chart.js/pull/4554. I was hoping it would be helpful to you in debugging, but it doesn't seem to work when I change `type: 'line'` to `type: 'bar'`

benmccann commented:
@nagix I want to thank you for the truly awesome description and all the time you've put into fixing this issue

I think there's one main issue, which is that it's making many other classes aware of the bar chart when they shouldn't need to know about the bar chart. It would be preferable to keep the fix contained within the bar chart as much as possible. If the bar chart looks at the ticks of the axis hopefully it should be possible to calculate the bounds of the bar

nagix commented:
@benmccann Thanks for reviewing this, and I understand your concern. To make it simpler, how about the following idea?  
- Don't use `sampleSize` option
- The default bar size is the size of the tick interval that the data sample is on
  - In case of a category scale, it will be a fixed size. In case of a time scale, it will depend on the tick configuration
- When detecting overlap with the next bars, recalculate the size individually so that they fit each other

benmccann commented:
That sounds reasonable to me. @simonbrunel thoughts?

Simon and I were talking about the one point below

> The default bar size is the size of the tick interval that the data

I think we came to the conclusion that it might be nice to have an option that switches this between two modes. One is the mode you mentioned. The other is a mode where you force the bars to be equally sized (the size of the closest two ticks). Here's a use case for what I'm talking about: imagine you're plotting a daily stock with a bar chart. You set `mode:linear`. I think you would want five equally sized bars with a space for the weekend followed by five equally sized bars for the next week. In this case it would be strange if the bars for Monday and Friday were much bigger than the bars for Tue, Wed, Thurs

benmccann commented:
Here's a sample that can be used to reproduce the bug I mentioned above when `mode: series` and `offsetGridLines: false`: https://raw.githubusercontent.com/benmccann/Chart.js/offsetGridlineExample/samples/scales/time/financial.html

benmccann commented:
Also, I haven't tested with your PR, but currently on master, if you try to plot a bar on a time scale, there's one point that doesn't get plotted (I'm not sure if it's the first or the last). That might be something else we want to test for

benmccann commented:
@nagix the latest commits from master are showing up in your PR now for some reason. I think you'll need to rebase

nagix commented:
I have changed the code so that the bar controller and the scales don't depend on each other as much as possible. Now bar controller calculates the bar width based on the return value of the `getPixelForValue` function.
- Remove `sampleSize` option
- Calculate the grid line offset based on the minimum interval between samples
- Add code to support `ticks.mode` and `ticks.source` option in the time scale
- Modify the interpolation table to support lookup for tick indexes
- Adjust the range of scales so that all bars can be displayed

nagix commented:
@benmccann The issue you reported has been fixed. Also, the last bar is plotted on a time scale, which was reported in #3297.

<img width="983" alt="screen shot 2017-07-28 at 6 29 50 am" src="https://user-images.githubusercontent.com/723188/28701934-f5c7090e-739c-11e7-967c-2bb2d4c7d1c4.png">

simonbrunel commented:
@nagix can you squash and rebase?

simonbrunel commented:
Thanks @nagix for this awesome work. I think there is still one (implementation) issue, can you post a screenshot of bars on a time scale with `mode: 'linear'`, `source: 'labels'`, with labels at irregular times:

![image](https://user-images.githubusercontent.com/3874900/28705318-7ce3fa4c-736f-11e7-958b-1a2b1ca8e93e.png)

benmccann commented:
@nagix will you be able to squash and rebase to get the latest commits from master to stop showing up in your PR? i'm very excited for this improvement to make its way in!

benmccann commented:
@nagix thanks again for this PR! is there anything we can do to help you figure out the rebase process?

nagix commented:
Sorry for the late response. I did squash and rebase to the latest master.

I have also added several tests to the time scale, category scale and bar controller to cover the `ticks.offset`, `barThickness` and `maxBarThickness` options.

When `barThickness` is not specified, the bar width will be sized based on the minimum interval between bars in the chart to avoid overlaps, and all bars will have the same width. When `'max'` is set to `barThickness`, the bar width will be maximized individually, and each bar can have a different width.

benmccann commented on docs/axes/cartesian/README.md:
```diff
@@ -31,6 +31,7 @@ The following options are common to all cartesian axes but do not apply to other
 | `maxRotation` | `Number` | `90` | Maximum rotation for tick labels when rotating to condense labels. Note: Rotation doesn't occur until necessary. *Note: Only applicable to horizontal scales.*
 | `minRotation` | `Number` | `0` | Minimum rotation for tick labels. *Note: Only applicable to horizontal scales.*
 | `mirror` | `Boolean` | `false` | Flips tick labels around axis, displaying the labels inside the chart instead of outside. *Note: Only applicable to vertical scales.*
+| `offset` | `Boolean` | `false` | If true, extra space is added to the both edges and the axis is scaled to fit into the chart area. This is set to `true` in the bar chart by default.
```
Sounds quite similar to the `scale.bounds` feature that was implemented in https://github.com/chartjs/Chart.js/pull/4556 (docs are being added in https://github.com/chartjs/Chart.js/pull/4592). Is it possible to use that option? I'm not quite sure if this is doing the same thing or not

benmccann replied:
Here's a couple screenshots to show the `scale.bounds` option

**`scale.bounds: 'data'`**

![bounds-data](https://user-images.githubusercontent.com/322311/28998830-78e898a0-79e9-11e7-9ad7-0e41c743cb7c.png)


**`scale.bounds: 'ticks'`**

![bounds-ticks](https://user-images.githubusercontent.com/322311/28998835-83ee4e84-79e9-11e7-8192-58c7a2917e00.png)

nagix replied:
When `scale.bounds: 'ticks'` but the first and last data points are on the ticks, the scale won't add extra spaces at edges (see 5 below). However, `offset: true ` will always add them to make sure that the whole bar rectangles are displayed (see 7 below).

<img width="410" alt="screen shot 2017-08-06 at 10 34 06 am" src="https://user-images.githubusercontent.com/723188/29000071-e398257c-7a92-11e7-93bb-67b0365cc1e6.png">
<img width="400" alt="screen shot 2017-08-06 at 10 34 13 am" src="https://user-images.githubusercontent.com/723188/29000072-ef7d0b00-7a92-11e7-8ae5-e12d67b7aabd.png">

Internally, `offset: true ` doesn't change the time/pos lookup table. It just add one half size of the next tick intervals in `getPixelForValue()` (or `getPixelForOffset()` in case of a time scale), and this can be easily applied to other scales such as a category scale.

benmccann commented:
Thanks @nagix!  There are some great changes here!

benmccann commented on src/controllers/controller.bar.js:
```diff
@@ -235,27 +245,76 @@ module.exports = function(Chart) {
+				curr = scale.getPixelForValue(null, i, datasetIndex);
+				data.push(curr);
+				minInterval = Math.min(minInterval, curr - prev);
+				if (ticksOffset) {
```
Do you need this if/else? Could you simplify by just always checking that `curr > min && curr < max`?

nagix replied:
We need to handle an edge case like that a dataset contains only one data point that is on the edge of a scale, the scale has only one tick and `ticks.offset: false`. In this case, we cannot calculate the bar width based on a tick interval or the distance from the nearest edge (because it is on the edge), so it uses the distance from the farthest edge.

andig commented:
Checking in here thanks to @etimberg. Amazing PR! Could you kindly share any fiddle that shows a combined bar/line chart, preferably with lines using x/y notation instead of labels?

nagix commented:
I have updated the code based on the discussion with @simonbrunel. I removed `barThickness: 'max'` option, and the bars are sized automatically so that they take available widths without overlap by default.

nagix commented:
@andig Does this help for you?
https://jsfiddle.net/nagix/657fxL8r/

etimberg commented on src/scales/scale.time.js:
```diff
@@ -562,6 +562,37 @@ module.exports = function(Chart) {
+
+			if (options.offset && ticks.length) {
+				if (!timeOpts.min) {
+					if (ticks.length > 1) {
```
To save a few bytes on size, I would update this to 

```javascript
upperValue = ticks.length > 1 ? ticks[1] : max
lowerValue = ticks[0]
me._leftOffset = (
  interpolate(me._table, 'time', upperValue, 'pos') -
  interpolate(me._table, 'time', lowerValue, 'pos')
) / 2;
```

nagix replied:
Agreed.

etimberg commented on src/scales/scale.time.js:
```diff
@@ -562,6 +562,37 @@ module.exports = function(Chart) {
+					}
+				}
+				if (!timeOpts.max) {
+					if (ticks.length > 1) {
```
similar comment here as above

etimberg commented on test/specs/controller.line.tests.js:
```diff
@@ -250,7 +250,7 @@ describe('Line controller tests', function() {
 		// 1 point
 		var point = meta.data[0];
-		expect(point._model.x).toBeCloseToPixel(262);
+		expect(point._model.x).toBeCloseToPixel(27);
```
is this change expected? Does it close any open issues? I seem to recall that some users wanted a single line point in the center of the chart. I would presume that adding `offset: true` still allows that use case. is that correct?

nagix replied:
Correct. If a user want a single line point in the center of the chart, one can use `offset: true`.

nagix commented on src/controllers/controller.bar.js:
```diff
@@ -236,26 +242,53 @@ module.exports = function(Chart) {
+			var pixels = [];
+			var ruler = [];
+			var isHorizontal = scale.isHorizontal();
+			var min = isHorizontal ? scale.left : scale.top;
```
Agreed.

nagix commented on src/controllers/controller.bar.js:
```diff
@@ -236,26 +242,53 @@ module.exports = function(Chart) {
+				}
+				leftCategorySize = leftSampleSize * options.categoryPercentage;
+				rightCategorySize = rightSampleSize * options.categoryPercentage;
+				fullBarSize = (leftCategorySize + rightCategorySize) / stackCount;
```
`leftCategorySize` is used a few lines below, so it doesn't seem to save bytes.

nagix commented on src/core/core.scale.js:
```diff
@@ -659,7 +661,7 @@ module.exports = function(Chart) {
 				var label = tick.label;
 				var lineWidth, lineColor, borderDash, borderDashOffset;
-				if (index === (typeof me.zeroLineIndex !== 'undefined' ? me.zeroLineIndex : 0)) {
+				if (index === (typeof me.zeroLineIndex !== 'undefined' ? me.zeroLineIndex : 0) && (options.offset === gridLines.offsetGridLines)) {
```
This prevents the zero line from being drawn in case of 2. and 3. below.
<img width="416" alt="screen shot 2017-08-10 at 11 37 10 am" src="https://user-images.githubusercontent.com/723188/29152114-873ae356-7dc0-11e7-9b49-f926b5662433.png">

nagix commented on src/core/core.scale.js:
```diff
@@ -693,8 +695,22 @@ module.exports = function(Chart) {
+					var xLineValue; // xvalues for grid lines
+					if (gridLines.offsetGridLines && ticks.length > 1) {
+						if (index === 0) {
+							xLineValue = (me.getPixelForTick(0) * 3 - me.getPixelForTick(1)) / 2;
```
When `gridLines.offsetGridLines` is true and a tick is the leftmost tick, grid lines move to the left by one half of the tick interval between itself and the next tick on the right.

nagix commented on src/scales/scale.category.js:
```diff
@@ -119,7 +121,7 @@ module.exports = function(Chart) {
 			}
 
-			return value;
+			return value + me.minIndex;
```
I don't think there is a ticket for it. `getValueForPixel()` is only called from test code.

simonbrunel commented on src/scales/scale.time.js:
```diff
@@ -562,6 +562,37 @@ module.exports = function(Chart) {
+			me._leftOffset = 0;
+			me._rightOffset = 0;
+
+			if (options.offset && ticks.length) {
```
I think the `buildTicks` method is already a bit too complex, can we move that logic in a separate external method:

```javascript
me._table = buildLookupTable(me._timestamps.data, min, max, options.distribution);
me._offsets = computeOffsets(table, ticks, options);   // return {left: ..., right: ...}

return ticksFromTimestamps(ticks, majorUnit);
```

simonbrunel replied:
BTW, did you try to compute offset based on `me._timestamps.data` instead of `ticks`?

simonbrunel replied:
Referring to this case (none linear data with linear ticks):

![image](https://user-images.githubusercontent.com/3874900/29162856-f0b266a2-7dba-11e7-9299-36c6e676f739.png)

nagix replied:
I haven't tested, but we cannot simply use `me._timestamps.data` to compute offsets because the array can contain timestamps from different types of datasets for example line with the point type and bars.

simonbrunel commented on src/scales/scale.time.js:
```diff
@@ -622,7 +653,7 @@ module.exports = function(Chart) {
 			var pos = interpolate(me._table, 'time', time, 'pos');
 
-			return start + size * pos;
+			return start + size * (me._leftOffset + pos) / (me._leftOffset + 1 + me._rightOffset);
```
Then would be `me._offsets.left` and `me._offsets.right`

simonbrunel commented on test/specs/controller.bar.tests.js:
```diff
@@ -1519,4 +1519,152 @@ describe('Bar controller tests', function() {
 		});
 	});
+
+	describe('Bar thickness with a category scale', function() {
```
Is there any way to generalize these 2 tests and avoid dependencies between the bar controller and category/time scales?

simonbrunel replied:
nm, will be reworked when we will refactor the scale logic.

simonbrunel commented:
@nagix might be good to update/simplify the PR description to keep it synced with the new implementation? Also, if you can list tickets for sure fixed by this PR at the end of the description, like in [this PR](https://github.com/chartjs/Chart.js/pull/4556) (with `Fixes #....` prefix), then tickets will be automatically closed as soon as this PR will be merged.

andig commented:
@nagix I've experimented somewhat with this PR and must say that I'm delighted. One thing I've noticed is that when I set `scale.time.unit: 'day'` I will still get gridlines at each x days at certain zoom levels. That is of course confusing. How should gridline spacing for bar charts over time series be handled- separate issue?

nagix commented on src/controllers/controller.bar.js:
```diff
@@ -154,7 +160,7 @@ module.exports = function(Chart) {
 			var base = vscale.getBasePixel();
 			var horizontal = vscale.isHorizontal();
-			var ruler = me._ruler || me.getRuler();
+			var ruler = (!me._ruler || index >= me._ruler.length) ? me.getRuler() : me._ruler;
```
Do you think it is a good idea to remove the ruler completely and to let `calculateBarIndexPixels()` calculate the bar size on the fly? That's because the bar size needs to be recalculated every time the position of the bar or the next bars is changed and there seems little reason to cache the recalculated information.

simonbrunel replied:
Good point, though the "first pass" should be cached.

nagix replied:
Ok, I will leave the ruler to cache `{start, end, pixels, stackCount, scale}` and let calculateBarIndexPixels() calculate the bar size using those values.

simonbrunel commented on src/core/core.scale.js:
```diff
@@ -693,8 +708,13 @@ module.exports = function(Chart) {
-					labelX = me.getPixelForTick(index, gridLines.offsetGridLines) + optionTicks.labelOffset; // x values for optionTicks (need to consider offsetLabel option)
+					var xLineValue = getLineValue(me, index, gridLines.offsetGridLines && ticks.length > 1);
+					if (xLineValue < me.left) {
+						lineColor = 'rgba(0,0,0,0)';
```
Why do we need to "hide" the grid line here?

nagix replied:
This prevents the leftmost grid line from being drawn outsize the chart area like this.

<img width="387" alt="screen shot 2017-08-12 at 3 35 45 am" src="https://user-images.githubusercontent.com/723188/29229146-6c97b678-7f0f-11e7-973d-fea7b3dd5baf.png">

simonbrunel replied:
Is it possible to skip that line instead? `if (xLineValue < me.left) { return; }`

nagix replied:
Since a grid line and a label are drawn together in the function, if we skip the line, the label will be also skipped ('1' in the chart above).

simonbrunel replied:
Right

nagix commented:
@andig For your first question, `time.unitStepSize: 1` will solve the problem. For your second question, `gridLines.offsetGridLines: true` is the solution with this PR if the tick interval is 1 day, and data points are at 00:00 every day. This option is set to true for a category scale but false for a time scale by default.

andig commented:
@nagix thanks for your feedback!

> time.unitStepSize: 1 will solve the problem. 

That has become `stepSize` now? Will check. Unfortunately overshadowed by https://github.com/chartjs/Chart.js/issues/4635 for me.

> For your second question, gridLines.offsetGridLines: true is the solution with this PR if the tick interval is 1 day, and data points are at 00:00 every day. This option is set to true for a category scale but false for a time scale by default.

Right. Unfortunately that is not sufficient for combines line/bar charts for physical units. If you think about electrical power (W) and daily consumption (Wh) it only makes sense to draw the bar (and the associated label) in the middle of the day. Otherwise the bar for daily consumption will overlap 2 days if plotted at 00:00 hours.

But that sounds like a separate issue that is not part of this PR. Thanks!

nagix commented:
@simonbrunel For #3747, this PR fixes a bar overlap issue as follows, but if it expects all bars to have one day width, it needs to wait for `barThickness: 'min'` or similar option as we discussed.
https://jsfiddle.net/nagix/vs8jve63/
https://jsfiddle.net/nagix/uLkn5jjc/

The same is true for #3780. If it expects the same bar width, this PR doesn't completely fix it. I removed these ticket numbers from the PR description just in case.
https://jsfiddle.net/nagix/53oejjav/1/

It fixes #3579 as you suggest.

nagix commented:
@andig You are right, `time.unitStepSize` was replaced with `time.stepSize`. I understand that `gridLines.offsetGridLines` doesn't fix your problem because lines and grid lines have to stay but bars and labels need to be shifted. How about moving bars by adding 12 hours and shifting labels by `labelOffset` option?

simonbrunel commented:
@nagix is this PR ready to be merged?

nagix commented:
@simonbrunel Yes, it is ready.

benmccann commented:
Woo! This is great! Thanks so much @nagix!

Will you be submitting a PR to add `barThickness: 'min'` next?

maxwell commented:
yay! Thank you!  Do you think this would warrant a new release?

myadav-etindia commented:
@nagix 
Just wanted to confirm if all the fixes done as part of this PR are available in chart.js release. We are using version 2.6.0 and that does not contain these fixes.
By when these fixes will be available ?

Thanks

andig commented:
@myadav-etindia this PR is in the master branch and will probably released with 2.7.0. See http://www.chartjs.org/docs/master/developers/ for how to get access to the development releases (= master branch).

myadav-etindia commented:
@andig 
Thanks. Got the latest.

myadav-etindia commented:
@andig 
Tentatively by when would the chart.js 2.7.0 release be available ?

andig commented:
@myadav-etindia I wouldn't know really- it's a question for the maintainers. My feeling is though that there is still a lot of movement that would not warrant a realease in the near future. You could still use a snapshot of the current master right now.

buddyp450 commented:
My team previously extended `elements.rectangle` because the existing Rectangle element assumed that the 'x' was the central point whose x<sub>1</sub> and x<sub>2</sub> were calculated dividing the width by 2. Our extended element assumes that x is the start and x + width = x<sub>2</sub>. I'll need to check this out to see if this is even needed anymore but I'm assuming it's not which would be a huge boon for us.

Offlein commented:
I was really pleased with the work done here, but one thing I'm confused about is the intention of how this handles time series line charts.

I have daily data that I'd like to present with labels grouped and labeled by month. That is, the gridLine should appear at the first of the month, but I'd expect the tick label (say "Sep" or "Jan") should appear centered between the two. It seems like with `offsetGridLines: true`, it appears visually the way one would expect, but the gridlines correspond to the halfway point in the month.

This fiddle illustrates what I mean: https://jsfiddle.net/y152r1nh/4/

Hovering over the point on the line before the label "Jan" gives me dates for "December 15". Am I misunderstanding how to use this feature, or is that an oversight?

benmccann commented:
This feature is moving the gridlines for bar charts. I.e. if you're drawing a bar chart and don't want the grid line through the middle of the bar then move the gridlines over. What's you're asking for is a feature probably more like `offsetLabels: true`

Offlein commented:
@benmccann Thanks. So that just plain doesn't exist, huh? I guess I'll look into how to implement such a thing.

flausinoth commented:
@nagix What if you have an interval of days? for example: from day 20 to day 23. When setting the values, the bar graph distorts.
